lib/blockchain/migrations.js

```diff
@@ -0,0 +1,342 @@
+    this.db = options.db;
+  }
+
+  /**
+   * Check if the migration applies to the database
+   * @returns {Promise}
+   */
+
+  async check() {
+    if (this.options.spv)
+      return types.FAKE_MIGRATE;
+
+    if (this.options.prune)
+      return types.SKIP;
+
+    return types.MIGRATE;
+  }
+
+  /**
+   * Log warnings when skipped.
+   */
+
+  warning() {
+    if (!this.options.prune)
+      throw new Error('No warnings to show!');
+
+    this.logger.warning('Pruned nodes cannot migrate the chain state.');
+    this.logger.warning('Your total chain value may be inaccurate!');
+  }
+
+  /**
+   * Actual migration
+   * @param {Batch} b
+   * @returns {Promise}
+   */
+
+  async migrate(b) {
+    this.logger.info('Migrating chain state.');
+    this.logger.info('This may take a few minutes...');
+
+    const state = await this.db.getState();
+    const tipHeight = await this.db.getHeight(state.tip);
+    const pending = state.clone();
+
+    pending.coin = 0;
+    pending.value = 0;
+    pending.burned = 0;
+
+    for (let height = 0; height <= tipHeight; height++) {
+      const hash = await this.db.getHash(height);
+      const block = await this.db.getBlock(hash);
+      assert(block);
+
+      const view = await this.db.getBlockView(block);
+
+      for (let i = 0; i < block.txs.length; i++) {
+        const tx = block.txs[i];
+
+        if (i > 0) {
+          for (const {prevout} of tx.inputs) {
+            const output = view.getOutput(prevout);
+            assert(output);
+
+            if (output.covenant.type >= rules.types.REGISTER
+                && output.covenant.type <= rules.types.REVOKE) {
+              continue;
+            }
+
+            pending.spend(output);
+          }
+        }
+
+        for (let i = 0; i < tx.outputs.length; i++) {
+          const output = tx.outputs[i];
+
+          if (output.isUnspendable())
+            continue;
+
+          if (output.covenant.isRegister())
+            pending.burn(output);
+
+          if (output.covenant.type >= rules.types.REGISTER
+              && output.covenant.type <= rules.types.REVOKE) {
+            continue;
+          }
+
+          if (output.covenant.isClaim()) {
+            if (output.covenant.getU32(5) !== 1)
+              continue;
+          }
+
+          pending.add(output);
+        }
+      }
+    }
+
+    b.put(layout.R.encode(), pending.encode());
+  }
+
+  static info() {
+    return {
+      name: 'Chain State migration',
+      description: 'Chain state is corrupted.'
+    };
+  }
+}
+
+/**
+ * Chain Migrator
+ * @alias module:blockchain.ChainMigrator
+ */
+class ChainMigrator extends Migrator {
+  /**
+   * Create ChainMigrator object.
+   * @constructor
+   * @param {Object} options
+   */
+
+  constructor(options) {
+    super(new ChainMigratorOptions(options));
+
+    this.logger = this.options.logger.context('chain-migrations');
+    this.flagError = 'Restart with `hsd --chain-migrate`.';
```

Comment:
If you like what I suggested [here](https://github.com/handshake-org/hsd/pull/615#issuecomment-889318944) then this error message could actually report the exact value (example `--chain-migrate=2`). Same goes for wallet.